File: apps/lifeos/taurireact-macapp/src/lib/services/beeper.ts
import { invoke } from "@tauri-apps/api/core";

// Types matching the Rust structs

export interface BeeperThread {
  thread_id: string;
  name: string;
  thread_type: string; // "dm" | "group"
  participant_count: number;
  message_count: number;
  last_message_at: string;
}

export interface BeeperMessage {
  message_id?: string;
  thread_id?: string;
  thread_name?: string;
  sender: string;
  text: string;
  timestamp_readable: string;
  timestamp?: number;
}

// Message for syncing to Convex
export interface BeeperSyncMessage {
  message_id: string;
  thread_id: string;
  sender: string;
  text: string;
  timestamp: number;
}

// Business thread marking (stored locally)
export interface BusinessThreadMark {
  threadId: string;
  isBusinessChat: boolean;
  businessNote?: string;
  markedAt: number;
}

export interface BeeperSyncResult {
  success: boolean;
  error?: string;
  message?: string;
}

export interface BeeperSyncProgress {
  status: "idle" | "checking" | "syncing" | "error" | "complete";
  currentStep: string;
  error?: string;
}

export const initialSyncProgress: BeeperSyncProgress = {
  status: "idle",
  currentStep: "",
};

// Check if running in Tauri
const isTauri = typeof window !== "undefined" && "__TAURI__" in window;

// ==================== SYNC INTERVAL SETTINGS ====================

export const SYNC_INTERVAL_OPTIONS = [
  { value: 1, label: "1 min" },
  { value: 3, label: "3 min" },
  { value: 5, label: "5 min" },
  { value: 10, label: "10 min" },
  { value: 15, label: "15 min" },
  { value: 30, label: "30 min" },
];

/**
 * Get the Beeper sync interval setting (default 3 minutes)
 */
export async function getBeeperSyncInterval(): Promise<number> {
  if (!isTauri) return 3;
  try {
    return await invoke<number>("get_beeper_sync_interval");
  } catch (error) {
    console.error("Failed to get Beeper sync interval:", error);
    return 3; // Default
  }
}

/**
 * Save the Beeper sync interval setting
 */
export async function saveBeeperSyncInterval(interval: number): Promise<void> {
  if (!isTauri) return;
  try {
    await invoke("save_beeper_sync_interval", { interval });
  } catch (error) {
    console.error("Failed to save Beeper sync interval:", error);
    throw error;
  }
}

/**
 * Check if Beeper is available (BeeperTexts folder exists)
 */
export async function checkBeeperAvailable(): Promise<boolean> {
  if (!isTauri) return false;
  try {
    return await invoke<boolean>("check_beeper_available");
  } catch (error) {
    console.error("Failed to check Beeper availability:", error);
    return false;
  }
}

/**
 * Check if the Beeper database has been synced (clean.duckdb exists)
 */
export async function checkBeeperDatabaseExists(): Promise<boolean> {
  if (!isTauri) return false;
  try {
    return await invoke<boolean>("check_beeper_database_exists");
  } catch (error) {
    console.error("Failed to check Beeper database:", error);
    return false;
  }
}

/**
 * Sync the Beeper database (runs pnpm sync && pnpm clean)
 */
export async function syncBeeperDatabase(): Promise<BeeperSyncResult> {
  if (!isTauri) {
    return { success: false, error: "Not running in Tauri" };
  }
  try {
    return await invoke<BeeperSyncResult>("sync_beeper_database");
  } catch (error) {
    console.error("Failed to sync Beeper database:", error);
    return { success: false, error: String(error) };
  }
}

/**
 * Get list of threads/conversations with optional search filter
 */
export async function getBeeperThreads(
  search?: string
): Promise<BeeperThread[]> {
  if (!isTauri) return [];
  try {
    return await invoke<BeeperThread[]>("get_beeper_threads", { search });
  } catch (error) {
    console.error("Failed to get Beeper threads:", error);
    return [];
  }
}

/**
 * Get conversation messages for a specific thread by exact name
 */
export async function getBeeperConversation(
  threadName: string,
  limit?: number
): Promise<BeeperMessage[]> {
  if (!isTauri) return [];
  try {
    return await invoke<BeeperMessage[]>("get_beeper_conversation", {
      threadName,
      limit,
    });
  } catch (error) {
    console.error("Failed to get Beeper conversation:", error);
    return [];
  }
}

/**
 * Get conversation messages for a specific thread by thread ID (preferred)
 * This avoids issues with duplicate thread names like "WhatsApp private chat"
 */
export async function getBeeperConversationById(
  threadId: string,
  limit?: number
): Promise<BeeperMessage[]> {
  if (!isTauri) return [];
  try {
    return await invoke<BeeperMessage[]>("get_beeper_conversation_by_id", {
      threadId,
      limit,
    });
  } catch (error) {
    console.error("Failed to get Beeper conversation by ID:", error);
    return [];
  }
}

/**
 * Get messages by contact/thread name (fuzzy search)
 */
export async function getBeeperMessages(
  name: string,
  limit?: number
): Promise<BeeperMessage[]> {
  if (!isTauri) return [];
  try {
    return await invoke<BeeperMessage[]>("get_beeper_messages", { name, limit });
  } catch (error) {
    console.error("Failed to get Beeper messages:", error);
    return [];
  }
}

/**
 * Search messages by text content
 */
export async function searchBeeperMessages(
  query: string
): Promise<BeeperMessage[]> {
  if (!isTauri) return [];
  try {
    return await invoke<BeeperMessage[]>("search_beeper_messages", { query });
  } catch (error) {
    console.error("Failed to search Beeper messages:", error);
    return [];
  }
}

/**
 * Get all messages for a thread (for syncing to Convex)
 */
export async function getBeeperMessagesForSync(
  threadId: string
): Promise<BeeperSyncMessage[]> {
  if (!isTauri) return [];
  try {
    return await invoke<BeeperSyncMessage[]>("get_beeper_messages_for_sync", {
      threadId,
    });
  } catch (error) {
    console.error("Failed to get Beeper messages for sync:", error);
    return [];
  }
}

// ==================== LOCAL STORAGE FOR BUSINESS MARKS ====================

const BUSINESS_MARKS_KEY = "beeper_business_marks";

/**
 * Get all business thread marks from localStorage
 */
export function getBusinessMarks(): Map<string, BusinessThreadMark> {
  if (typeof window === "undefined") return new Map();
  try {
    const stored = localStorage.getItem(BUSINESS_MARKS_KEY);
    if (!stored) return new Map();
    const marks = JSON.parse(stored) as BusinessThreadMark[];
    return new Map(marks.map((m) => [m.threadId, m]));
  } catch (error) {
    console.error("Failed to load business marks:", error);
    return new Map();
  }
}

/**
 * Save business thread marks to localStorage
 */
function saveBusinessMarks(marks: Map<string, BusinessThreadMark>): void {
  if (typeof window === "undefined") return;
  try {
    const marksArray = Array.from(marks.values());
    localStorage.setItem(BUSINESS_MARKS_KEY, JSON.stringify(marksArray));
  } catch (error) {
    console.error("Failed to save business marks:", error);
  }
}

/**
 * Mark a thread as business (or unmark)
 */
export function markThreadAsBusiness(
  threadId: string,
  isBusinessChat: boolean,
  businessNote?: string
): void {
  const marks = getBusinessMarks();
  if (isBusinessChat) {
    marks.set(threadId, {
      threadId,
      isBusinessChat: true,
      businessNote,
      markedAt: Date.now(),
    });
  } else {
    marks.delete(threadId);
  }
  saveBusinessMarks(marks);
}

/**
 * Check if a thread is marked as business
 */
export function isThreadMarkedAsBusiness(threadId: string): boolean {
  const marks = getBusinessMarks();
  return marks.has(threadId);
}

/**
 * Get business note for a thread
 */
export function getBusinessNote(threadId: string): string | undefined {
  const marks = getBusinessMarks();
  return marks.get(threadId)?.businessNote;
}

/**
 * Get all business thread IDs
 */
export function getBusinessThreadIds(): string[] {
  const marks = getBusinessMarks();
  return Array.from(marks.keys());
}

File: apps/lifeos/taurireact-macapp/src-tauri/src/beeper.rs
use serde::{Deserialize, Serialize};
use std::path::PathBuf;
use std::process::Command;
use tauri::command;

/// Represents a Beeper thread/conversation
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct BeeperThread {
    pub name: String,
    #[serde(rename = "type")]
    pub thread_type: String,
    pub participant_count: i64,
    pub message_count: i64,
    pub last_message_at: String,
}

/// Represents a Beeper message
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct BeeperMessage {
    #[serde(default)]
    pub thread_name: Option<String>,
    pub sender: String,
    pub text: String,
    pub timestamp_readable: String,
}

/// Result of syncing the Beeper database
#[derive(Debug, Serialize, Deserialize)]
pub struct BeeperSyncResult {
    pub success: bool,
    pub error: Option<String>,
    pub message: Option<String>,
}

/// Get the path to the beeperdb package
fn get_beeperdb_path() -> PathBuf {
    // Navigate from src-tauri to the monorepo root, then to packages/beeperdb
    let current_dir = std::env::current_dir().unwrap_or_default();

    // If we're in src-tauri during development
    if current_dir.ends_with("src-tauri") {
        return current_dir
            .parent() // taurireact-macapp
            .and_then(|p| p.parent()) // lifeos
            .and_then(|p| p.parent()) // apps
            .and_then(|p| p.parent()) // monorepo root
            .map(|p| p.join("packages").join("beeperdb"))
            .unwrap_or_else(|| PathBuf::from("packages/beeperdb"));
    }

    // If we're in the app directory during development
    if current_dir.ends_with("taurireact-macapp") {
        return current_dir
            .parent() // lifeos
            .and_then(|p| p.parent()) // apps
            .and_then(|p| p.parent()) // monorepo root
            .map(|p| p.join("packages").join("beeperdb"))
            .unwrap_or_else(|| PathBuf::from("packages/beeperdb"));
    }

    // Try to find monorepo root by looking for package.json with "hola" workspace
    let mut search_dir = current_dir.clone();
    for _ in 0..10 {
        let packages_dir = search_dir.join("packages").join("beeperdb");
        if packages_dir.exists() {
            return packages_dir;
        }
        if let Some(parent) = search_dir.parent() {
            search_dir = parent.to_path_buf();
        } else {
            break;
        }
    }

    // Fallback: assume relative to home directory
    dirs::home_dir()
        .map(|h| {
            h.join("Sync")
                .join("00.Projects")
                .join("holaai-convexo-monorepo")
                .join("packages")
                .join("beeperdb")
        })
        .unwrap_or_else(|| PathBuf::from("packages/beeperdb"))
}

/// Check if BeeperTexts folder exists (Beeper is available)
#[command]
pub async fn check_beeper_available() -> Result<bool, String> {
    let beeper_path = dirs::home_dir()
        .map(|h| {
            h.join("Library")
                .join("Application Support")
                .join("BeeperTexts")
        })
        .unwrap_or_default();

    Ok(beeper_path.exists())
}

/// Check if clean.duckdb exists (data has been synced)
#[command]
pub async fn check_beeper_database_exists() -> Result<bool, String> {
    let beeperdb_path = get_beeperdb_path();
    let db_path = beeperdb_path.join("data").join("clean.duckdb");
    Ok(db_path.exists())
}

/// Sync the Beeper database by running pnpm sync && pnpm clean
#[command]
pub async fn sync_beeper_database() -> Result<BeeperSyncResult, String> {
    let beeperdb_path = get_beeperdb_path();

    if !beeperdb_path.exists() {
        return Ok(BeeperSyncResult {
            success: false,
            error: Some(format!(
                "beeperdb package not found at: {}",
                beeperdb_path.display()
            )),
            message: None,
        });
    }

    // Run pnpm sync (clone DBs + build export.sqlite)
    let sync_output = Command::new("pnpm")
        .args(["sync"])
        .current_dir(&beeperdb_path)
        .output()
        .map_err(|e| format!("Failed to run pnpm sync: {}", e))?;

    if !sync_output.status.success() {
        let stderr = String::from_utf8_lossy(&sync_output.stderr);
        return Ok(BeeperSyncResult {
            success: false,
            error: Some(format!("pnpm sync failed: {}", stderr)),
            message: None,
        });
    }

    // Run pnpm clean (build clean.duckdb)
    let clean_output = Command::new("pnpm")
        .args(["clean"])
        .current_dir(&beeperdb_path)
        .output()
        .map_err(|e| format!("Failed to run pnpm clean: {}", e))?;

    if !clean_output.status.success() {
        let stderr = String::from_utf8_lossy(&clean_output.stderr);
        return Ok(BeeperSyncResult {
            success: false,
            error: Some(format!("pnpm clean failed: {}", stderr)),
            message: None,
        });
    }

    Ok(BeeperSyncResult {
        success: true,
        error: None,
        message: Some("Beeper database synced successfully".to_string()),
    })
}

/// Get list of threads/conversations
#[command]
pub async fn get_beeper_threads(search: Option<String>) -> Result<Vec<BeeperThread>, String> {
    let beeperdb_path = get_beeperdb_path();

    // Build the command: bun query.ts threads [search]
    let mut cmd = Command::new("bun");
    cmd.arg("query.ts").arg("threads");

    if let Some(ref s) = search {
        cmd.arg(s);
    }

    let output = cmd
        .current_dir(&beeperdb_path)
        .output()
        .map_err(|e| format!("Failed to run bun query.ts threads: {}", e))?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(format!("bun query.ts threads failed: {}", stderr));
    }

    let stdout = String::from_utf8_lossy(&output.stdout);

    // Parse JSON output
    let threads: Vec<BeeperThread> =
        serde_json::from_str(&stdout).map_err(|e| format!("Failed to parse threads JSON: {}", e))?;

    Ok(threads)
}

/// Get conversation/messages for a specific thread by exact name
#[command]
pub async fn get_beeper_conversation(
    thread_name: String,
    limit: Option<i32>,
) -> Result<Vec<BeeperMessage>, String> {
    let beeperdb_path = get_beeperdb_path();

    // Build the command: bun query.ts convo "<thread_name>"
    let mut cmd = Command::new("bun");
    cmd.arg("query.ts").arg("convo").arg(&thread_name);

    if let Some(l) = limit {
        cmd.arg(l.to_string());
    }

    let output = cmd
        .current_dir(&beeperdb_path)
        .output()
        .map_err(|e| format!("Failed to run bun query.ts convo: {}", e))?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(format!("bun query.ts convo failed: {}", stderr));
    }

    let stdout = String::from_utf8_lossy(&output.stdout);

    // Parse JSON output
    let messages: Vec<BeeperMessage> = serde_json::from_str(&stdout)
        .map_err(|e| format!("Failed to parse conversation JSON: {}", e))?;

    Ok(messages)
}

/// Search messages by text content
#[command]
pub async fn search_beeper_messages(query: String) -> Result<Vec<BeeperMessage>, String> {
    let beeperdb_path = get_beeperdb_path();

    // Build the command: bun query.ts search "<query>"
    let output = Command::new("bun")
        .arg("query.ts")
        .arg("search")
        .arg(&query)
        .current_dir(&beeperdb_path)
        .output()
        .map_err(|e| format!("Failed to run bun query.ts search: {}", e))?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(format!("bun query.ts search failed: {}", stderr));
    }

    let stdout = String::from_utf8_lossy(&output.stdout);

    // Parse JSON output
    let messages: Vec<BeeperMessage> = serde_json::from_str(&stdout)
        .map_err(|e| format!("Failed to parse search results JSON: {}", e))?;

    Ok(messages)
}

/// Get messages by contact/thread name (fuzzy search)
#[command]
pub async fn get_beeper_messages(
    name: String,
    limit: Option<i32>,
) -> Result<Vec<BeeperMessage>, String> {
    let beeperdb_path = get_beeperdb_path();

    // Build the command: bun query.ts messages "<name>" [limit]
    let mut cmd = Command::new("bun");
    cmd.arg("query.ts").arg("messages").arg(&name);

    if let Some(l) = limit {
        cmd.arg(l.to_string());
    }

    let output = cmd
        .current_dir(&beeperdb_path)
        .output()
        .map_err(|e| format!("Failed to run bun query.ts messages: {}", e))?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(format!("bun query.ts messages failed: {}", stderr));
    }

    let stdout = String::from_utf8_lossy(&output.stdout);

    // Parse JSON output
    let messages: Vec<BeeperMessage> = serde_json::from_str(&stdout)
        .map_err(|e| format!("Failed to parse messages JSON: {}", e))?;

    Ok(messages)
}
